File: api/src/models/lyrics.rs
use crate::{models::generic::Pageable, utils::de::de_opt_i64};
use serde::{Deserialize, Serialize};
use sqlx::FromRow;

#[derive(Debug, Clone, Serialize, FromRow)]
#[serde(rename_all = "camelCase")]
pub struct Lyrics {
    pub id: i64,
    pub synced: bool,
    pub file_path: String,
    pub checksum: String,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct LyricsContent {
    pub text: String,
}

pub struct CreateLyrics {
    pub language: Option<String>,
    pub provider: Option<String>,
    pub synced: bool,
    pub file_path: String,
    pub checksum: String,
    pub track_id: i64,
}

#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct LyricsFilters {
    #[serde(default, deserialize_with = "de_opt_i64")]
    pub artist_id: Option<i64>,
    #[serde(default, deserialize_with = "de_opt_i64")]
    pub album_id: Option<i64>,
    #[serde(default, deserialize_with = "de_opt_i64")]
    pub track_id: Option<i64>,
}

#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct LyricsQuery {
    #[serde(flatten)]
    pub pageable: Pageable,
    #[serde(flatten)]
    pub filters: LyricsFilters,
}

File: web/src/domain/lyrics.ts
import z from "zod";
import { pageableSchema } from "./generic";

export interface Lyrics {
  id: number;
  synced: boolean;
  filePath: string;
  checksum: string;
}

export interface LyricsContent {
  text: string;
}

export const lyricsSearchSchema = pageableSchema.extend({
  artistId: z.int().min(1).optional().catch(undefined),
  albumId: z.int().min(1).optional().catch(undefined),
  trackId: z.int().min(1).optional().catch(undefined),
});

export type LyricsSearch = z.infer<typeof lyricsSearchSchema>;
